draft-ietf-quic-tls.md

```diff
@@ -580,6 +661,185 @@ Prior to TLS providing keys, no record protection is performed and the
+valid session resumption ticket, or because the client does not wish to send any
+0-RTT data.
+
+If the client does not have 0-RTT data to send, it transitions to using 1-RTT
+keys before sending its second flight of TLS handshake messages (that is, the
+flight containing the TLS Finished message).  Subsequent messages are protected
+with 1-RTT keys and have a KEY_PHASE bit set to 1.
+
+If the client sends 0-RTT data, it marks packets protected with 0-RTT keys with
+a KEY_PHASE of 1.  The later transition to using 1-RTT keys then results in the
+client using a KEY_PHASE of 0 for packets that are protected with 1-RTT keys.
+
+The server transitions to using 1-RTT keys after sending its first flight of TLS
+handshake messages.  From this point, the server protects all packets with 1-RTT
+keys.  Future packets are therefore protected with 1-RTT keys and marked with a
+KEY_PHASE of 1.
+
+Like the client, a server MUST send retransmissions of its unprotected handshake
+messages or acknowledgments for unprotected handshake messages sent by the
+client in unprotected packets (KEY_PHASE=0).
+
+The second flight of TLS handshake messages from a client is always protected
+with the current packet protection keys.  This includes the TLS
+end_of_early_data alert, which a client sends before its final set of handshake
+messages.
+
+
+### Retransmission and Acknowledgment of Unprotected Packets
+
+The handshake messages the first flight of handshake messages from both client
+and server (ClientHello, or ServerHello through to the server's Finished) a
+critical to the key exchange.  The contents of these messages is used to
+determine the keys used to protect later messages.  If these are protected with
+newer keys, they will be indecipherable to the recipient.
+
+Even though newer keys could be available, the first flight of TLS handshake
+messages sent by an endpoint MUST NOT be sent in protected packets,
+specifically:
+
+* A client MUST NOT restransmit a TLS ClientHello with 0-RTT keys.  The server
+  needs this message in order to determine the 0-RTT keys.
+
+* A server MUST NOT retransmit any of its TLS handshake messages with 1-RTT
+  keys.  The client needs these messages in order to determine the 1-RTT keys.
+
+A HelloRetryRequest might be used to reject an initial ClientHello.  A
+HelloRetryRequest handshake message and any second ClientHello that is sent in
+response MUST also be sent without packet protection.  This is natural, because
+no new keying material will be available when these messages need to be sent.
+
+Note:
+
+: An alternative way of identifying handshake data that needs to be sent without
+  protection is to collect all handshake data from before TLS provides the first
+  keys (see {{key-ready-events}}).
+
+Retransmissions of these handshake messages MUST be send in unprotected packets
+(with a KEY_PHASE of 0).  Any ACK frames for these messages MUST also be send in
```

Comment:
The problem we have is that an ACK might not be readable if it is encrypted.  See #34.

1. If the server ACKs the ClientHello by encrypting it, that's probably OK, it has to do that with 1-RTT keys.  But the client might not have the 1-RTT keys because it needs the entire set of server handshake messages to that.

2. If the client ACKs the ServerHello (et al.) with 0-RTT keys, then those ACKs are lost if the server rejects 0-RTT.  I guess you could say that this is OK on the basis that we lose ACKs too, but I think that it could force the server into retransmission on a timer (ugh).

3. If the client ACKs the ServerHello (et al.) with 1-RTT keys, then the server should be able to decrypt them perfectly well.  However, the only value gained from this is better feedback on timing, it doesn't help with loss recovery because there was none.